src/material-experimental/mdc-button-toggle/button-toggle.ts

```diff
@@ -0,0 +1,672 @@
+  constructor(
+    /** MatButtonToggle that emits the event. */
+    public source: MatButtonToggle,
+    /** Value assigned to the MatButtonToggle. */
+    public value: any) {
+  }
+}
+
+export const _MatButtonToggleBaseMixin: CanDisableRippleCtor & CanDisableCtor & CanColorCtor &
+  typeof MatButtonToggleMixinCore =
+  mixinColor(mixinDisabled(mixinDisableRipple(MatButtonToggleMixinCore)));
+
+/** Exclusive selection button toggle group that behaves like a radio-button group. */
+@Directive({
+  selector: 'mat-button-toggle-group',
+  providers: [
+    MAT_BUTTON_TOGGLE_GROUP_VALUE_ACCESSOR,
+    {provide: MAT_BUTTON_TOGGLE_GROUP, useExisting: MatButtonToggleGroup},
+  ],
+  host: {
+    '[attr.role]': 'multiple ? "group" : "radiogroup"',
+    'class': 'mat-mdc-button-toggle-group mdc-segmented-button',
+    '[attr.aria-disabled]': 'disabled',
+    '[class.mdc-segmented-button__single-select]': '!multiple',
+  },
+  exportAs: 'matButtonToggleGroup',
+})
+export class MatButtonToggleGroup implements ControlValueAccessor, OnInit, AfterViewInit,
+  AfterContentInit, OnDestroy {
+  /**
+   * Reference to the raw value that the consumer tried to assign. The real
+   * value will exclude any values from this one that don't correspond to a
+   * toggle. Useful for the cases where the value is assigned before the toggles
+   * have been initialized or at the same that they're being swapped out.
+   */
+  private _rawValue: any;
+
+  /**
+   * The method to be called in order to update ngModel.
+   * Now `ngModel` binding is not supported in multiple selection mode.
+   */
+  _controlValueAccessorChangeFn: (value: any) => void = () => {};
+
+  /** onTouch function registered via registerOnTouch (ControlValueAccessor). */
+  _onTouched: () => any = () => {};
+
+  /** Child button toggle buttons. */
+  @ContentChildren(forwardRef(() => MatButtonToggle), {
+    // Note that this would technically pick up toggles
+    // from nested groups, but that's not a case that we support.
+    descendants: true
+  }) _buttonToggles: QueryList<MatButtonToggle>;
+
+  _foundation: MDCSegmentedButtonFoundation;
+
+  private _selectionModel: SelectionModel<MatButtonToggle>;
+
+  private _adapter: MDCSegmentedButtonAdapter = {
+    hasClass: (name: string) => this._elementRef.nativeElement.classList.contains(name),
+    getSegments: () => {
+      return this._buttonToggles.map(segment => {
+        return {index: segment.index, selected: segment.checked} as SegmentDetail;
+      });
+    },
+    selectSegment: (indexOrSegmentId: number | string) => this._selectValue(indexOrSegmentId),
+    unselectSegment: (indexOrSegmentId: number | string) => this._unselectValue(indexOrSegmentId),
+    notifySelectedChange: (detail: SegmentDetail) => this._emitChangeEvent(),
+  };
+
+  constructor(private _changeDetector: ChangeDetectorRef,
+              private _elementRef: ElementRef<HTMLElement>) {
+    this._foundation = new MDCSegmentedButtonFoundation(this._adapter);
```

Comment:
I notice we're not using many of the foundation methods: `selectSegment`, `unselectSegment`, `handleSelected`, `isSegmentSelected`, `getSelectedSegments`. Should we be? If not, what exactly are we getting out of using the foundation?

Reply:
I'm a little confused on how to incorporate the foundation methods as the inputs and outputs are of `SegmentDetail` and I think the original API deals directly with `MatButtonToggle` components etc

Reply:
It looks like `SegmentDetail` is just the index an selected state. You could turn a list of `ButtonToggle` into a list of `SegmentDetail` like this:

```
toggles.toArray().map((t, i) => ({index: i, selected: t.selected}))
```

And you can convert a `SegmentDetail` to a `ButtonToggle` like this:

```
toggles.toArray()[detail.index]
```